src/SignalR/clients/cpp/src/signalrclient/hub_connection_impl.cpp

```diff
@@ -96,67 +96,89 @@ namespace signalr
-                        previous_task.get();
-                        return pplx::task<void>(connection->m_handshakeTask);
-                    })
-                    .then([weak_connection](pplx::task<void> previous_task)
+
+                if (start_exception)
+                {
+                    connection->m_connection->stop([start_exception, callback, weak_connection](std::exception_ptr)
                     {
                         try
-                        {
-                            previous_task.get();
-                            return previous_task;
-                        }
-                        catch (std::exception e)
                         {
                             auto connection = weak_connection.lock();
-                            if (connection)
+                            if (!connection)
                             {
-                                return connection->m_connection->stop()
-                                    .then([e]() {
-                                        throw e;
-                                    });
+                                callback(std::make_exception_ptr(signalr_exception("the hub connection has been deconstructed")));
+                                return;
                             }
-                            throw e;
+                            pplx::task<void>(connection->m_handshakeTask).get();
                         }
+                        catch (...) {}
+
+                        callback(start_exception);
                     });
+                    return;
+                }
+
+                // TODO: Generate this later when we have the protocol abstraction
+                auto handshake_request = "{\"protocol\":\"json\",\"version\":1}\x1e";
+                connection->m_connection->send(handshake_request, [weak_connection, callback](std::exception_ptr exception)
+                {
+                    auto connection = weak_connection.lock();
+                    if (!connection)
+                    {
+                        // The connection has been destructed
+                        callback(std::make_exception_ptr(signalr_exception("the hub connection has been deconstructed")));
+                        return;
+                    }
+
+                    if (exception)
+                    {
+                        callback(exception);
+                        return;
+                    }
+
+                    try
+                    {
+                        pplx::task<void>(connection->m_handshakeTask).get();
+                        callback(nullptr);
+                    }
+                    catch (...)
+                    {
+                        auto handshake_exception = std::current_exception();
+                        connection->m_connection->stop([callback, handshake_exception](std::exception_ptr)
+                        {
+                            callback(handshake_exception);
```

Comment:
Can we merge or log the stop exception if there is one?

Reply:
We log the exception in connection_impl::stop